File: app/src/main/java/com/github/malitsplus/shizurunotes/ui/charalist/CharaListFragment.java
package com.github.malitsplus.shizurunotes.ui.charalist;

import android.content.Context;
import android.icu.text.Transliterator;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.core.view.GravityCompat;
import androidx.databinding.DataBindingUtil;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.github.malitsplus.shizurunotes.common.Statics;
import com.github.malitsplus.shizurunotes.ui.MainActivity;
import com.github.malitsplus.shizurunotes.R;
import com.github.malitsplus.shizurunotes.common.UpdateManager;
import com.github.malitsplus.shizurunotes.databinding.FragmentCharaBinding;
import com.github.malitsplus.shizurunotes.ui.SharedViewModel;
import com.github.malitsplus.shizurunotes.ui.SharedViewModelFactory;

public class CharaListFragment extends Fragment {

    private SharedViewModel sharedViewModel;
    private CharaListViewModel charaListViewModel;
    private CharaListAdapter adapter;
    private LinearLayoutManager layoutManager;
    private RecyclerView recyclerView;
    private DrawerLayout drawerLayout;

    @Override
    public void onAttach(@NonNull Context context){
        super.onAttach(context);
        //sometimes will occur NPE?
        //((MainActivity)context).updateManager.setIFragmentCallBack(this);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container,
                             Bundle savedInstanceState) {

        sharedViewModel = new ViewModelProvider(getActivity()).get(SharedViewModel.class);
        SharedViewModelFactory factory = new SharedViewModelFactory(sharedViewModel);
        charaListViewModel = new ViewModelProvider(this, factory).get(CharaListViewModel.class);

        FragmentCharaBinding binding = DataBindingUtil.inflate(inflater, R.layout.fragment_chara, container, false);
        binding.setViewModel(charaListViewModel);
        binding.setLifecycleOwner(this);

        //get controls
        drawerLayout = binding.charaDrawer;
        recyclerView = binding.charaListRecycler;

        //set layout manager
        layoutManager = new LinearLayoutManager(getActivity());
        recyclerView.setLayoutManager(layoutManager);

        //set adapter
        adapter = new CharaListAdapter(this.getContext(), sharedViewModel);
        recyclerView.setAdapter(adapter);

        //减少开销
        recyclerView.setHasFixedSize(true);
        recyclerView.setItemViewCacheSize(200);

        //设置观察者
        charaListViewModel.getLiveCharaList().observe(getViewLifecycleOwner(), (charaList) ->
            adapter.update(charaList)
        );

        setHasOptionsMenu(true);
        setButtonListener(binding);
        return binding.getRoot();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater){
        inflater.inflate(R.menu.fragment_chara_bar, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item){
        if(item.getItemId() == R.id.action_filter) {
            if (!drawerLayout.isDrawerOpen(GravityCompat.END)) {
                drawerLayout.openDrawer(GravityCompat.END);
            } else {
                drawerLayout.closeDrawer(GravityCompat.END);
            }

        }
        return false;
    }

    private void setButtonListener(FragmentCharaBinding binding){
        binding.btnConfirm.setOnClickListener((v) ->{

            String position;
            switch (binding.chipGroupPosition.getCheckedChipId()){
                case R.id.chip_position_forward :
                    position = Statics.FILTER_FORWARD;
                    break;
                case R.id.chip_position_middle :
                    position = Statics.FILTER_MIDDLE;
                    break;
                case R.id.chip_position_rear :
                    position = Statics.FILTER_REAR;
                    break;
                default :
                    position = Statics.FILTER_NULL;
                    break;
            }

            int type;
            switch (binding.chipGroupAtkType.getCheckedChipId()){
                case R.id.chip_type_physical :
                    type = 1;
                    break;
                case R.id.chip_type_magical :
                    type = 2;
                    break;
                default :
                    type = 0;
                    break;
            }

            CharaListViewModel.SortValue sortValue;
            switch (binding.chipGroupSort.getCheckedChipId()){
                case R.id.chip_sort_position :
                    sortValue = CharaListViewModel.SortValue.POSITION;
                    break;
                case R.id.chip_sort_physical_atk :
                    sortValue = CharaListViewModel.SortValue.ATK;
                    break;
                case R.id.chip_sort_magical_atk :
                    sortValue = CharaListViewModel.SortValue.MAGIC_ATK;
                    break;
                case R.id.chip_sort_physical_def :
                    sortValue = CharaListViewModel.SortValue.DEF;
                    break;
                case R.id.chip_sort_magical_def :
                    sortValue = CharaListViewModel.SortValue.MAGIC_DEF;
                    break;
                case R.id.chip_sort_age :
                    sortValue = CharaListViewModel.SortValue.AGE;
                    break;
                case R.id.chip_sort_height :
                    sortValue = CharaListViewModel.SortValue.HEIGHT;
                    break;
                case R.id.chip_sort_weight :
                    sortValue = CharaListViewModel.SortValue.WEIGHT;
                    break;
                case R.id.chip_sort_burst_size:
                    sortValue = CharaListViewModel.SortValue.BUST_SIZE;
                    break;
                default :
                    sortValue = CharaListViewModel.SortValue.NEW;
                    break;
            }

            boolean asc;
            if(binding.chipGroupSortWay.getCheckedChipId() == R.id.chip_asc)
                asc = true;
            else
                asc = false;
            charaListViewModel.filter(position, type, sortValue, asc);
        });
    }

    public void updateList(){
        charaListViewModel.filterDefault();
    }
}

File: app/src/main/java/com/github/malitsplus/shizurunotes/ui/charalist/CharaListFragment.kt
package com.github.malitsplus.shizurunotes.ui.charalist

import android.content.Context
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.fragment.FragmentNavigator
import androidx.navigation.fragment.FragmentNavigatorExtras
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import com.github.malitsplus.shizurunotes.R
import com.github.malitsplus.shizurunotes.data.Chara
import com.github.malitsplus.shizurunotes.databinding.FragmentCharaListBinding
import com.github.malitsplus.shizurunotes.databinding.ItemCharaBinding
import com.github.malitsplus.shizurunotes.ui.BottomNaviFragmentDirections
import com.github.malitsplus.shizurunotes.ui.base.*
import com.github.malitsplus.shizurunotes.ui.shared.SharedViewModelChara
import com.github.malitsplus.shizurunotes.ui.shared.SharedViewModelCharaFactory
import com.github.malitsplus.shizurunotes.ui.shared.SharedViewModelEquipment
import org.w3c.dom.Text


class CharaListFragment : Fragment(), OnCharaActionListener
{
    private lateinit var binding: FragmentCharaListBinding
    private lateinit var charaListVM: CharaListViewModel
    private lateinit var sharedChara: SharedViewModelChara
    private lateinit var sharedEquipment: SharedViewModelEquipment
    private val charaListAdapter by lazy { ViewTypeAdapter<ViewType<*>>(onItemActionListener = this) }

    override fun onResume() {
        super.onResume()
        binding.apply {
            dropdownText1.dismissDropDown()
            dropdownText2.dismissDropDown()
            dropdownText3.dismissDropDown()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        sharedChara = ViewModelProvider(requireActivity())[SharedViewModelChara::class.java]
        sharedEquipment = ViewModelProvider(requireActivity())[SharedViewModelEquipment::class.java]
        charaListVM = ViewModelProvider(this, SharedViewModelCharaFactory(sharedChara))[CharaListViewModel::class.java]
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentCharaListBinding.inflate(inflater, container, false)
        setDropdownText()
        setObserver()
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        with (binding) {
            charaSearchBar.addTextChangeListener(charaListVM.textWatcher)
            charaListAdapter.setList(charaListVM.getViewList())
            charaListRecycler.apply {
                adapter = charaListAdapter
                layoutManager = LinearLayoutManager(requireContext())
                setHasFixedSize(true)

                parentFragment?.postponeEnterTransition()
                viewTreeObserver.addOnPreDrawListener {
                    parentFragment?.startPostponedEnterTransition()
                    true
                }
            }
        }
    }

    private fun setHintTextVisibility(downloadDb: Boolean = false, noResult: Boolean = false) {
        binding.downloadDbHint.visibility = if (downloadDb) View.VISIBLE else View.GONE
        binding.textSearchNoResult.visibility = if (noResult) View.VISIBLE else View.GONE
    }

    private fun setObserver() {
        charaListVM.liveCharaList.observe(viewLifecycleOwner, Observer {
            if (sharedChara.charaList.value.isNullOrEmpty() && sharedChara.loadingFlag.value == false && sharedEquipment.loadingFlag.value == false) {
                setHintTextVisibility(downloadDb = true)
            } else if (sharedChara.charaList.value?.isNotEmpty() == true && it.isEmpty()) {
                setHintTextVisibility(noResult = true)
            } else {
                setHintTextVisibility()
            }
            charaListAdapter.setUpdatedList(charaListVM.getViewList(it))
        })

        sharedChara.loadingFlag.observe(viewLifecycleOwner, Observer {
            binding.charaListProgressBar.visibility = if (it) {
                View.VISIBLE
            } else {
                View.GONE
            }
        })

        sharedEquipment.loadingFlag.observe(viewLifecycleOwner, Observer {
            if (it) {
                binding.charaListProgressBar.visibility = View.VISIBLE
            }
        })

        sharedChara.charaList.observe(viewLifecycleOwner, Observer {
            updateList()
        })
    }

    private fun setDropdownText(){
        binding.apply {
            dropdownText1.apply {
                onItemClickListener = charaListVM.getSpinnerClickListener(1)
                setAdapter(getSpinnerAdapter(1))
            }
            dropdownText2.apply {
                onItemClickListener = charaListVM.getSpinnerClickListener(2)
                setAdapter(getSpinnerAdapter(2))
            }
            dropdownText3.apply {
                onItemClickListener = charaListVM.getSpinnerClickListener(3)
                setAdapter(getSpinnerAdapter(3))
            }
        }
    }

    private fun updateList() {
        charaListVM.filterDefault()
    }

    private fun getSpinnerAdapter(type: Int): MaterialSpinnerAdapter<String> {
        return MaterialSpinnerAdapter(
            requireContext(),
            R.layout.dropdown_item_chara_list,
            charaListVM.dropDownValuesMap[type] ?: arrayOf()
        )
    }

    override fun onCharaClickedListener(chara: Chara, position: Int) {
        sharedChara.mSetSelectedChara(chara)
        sharedChara.backFlag = false
        val viewHolder = binding.charaListRecycler.findViewHolderForAdapterPosition(position) as ViewTypeHolder.GeneralViewHolder
        val extras = FragmentNavigatorExtras(
            (viewHolder.binding as ItemCharaBinding).itemChara to "transItem_${chara.charaId}"
        )
        val action = BottomNaviFragmentDirections.actionNavBottomNavigationToNavCharaDetails().setCharaId(chara.charaId)
        findNavController().navigate(action, extras)
    }

    override fun onItemClicked(position: Int) {
    }
}